### ISSUE TITLE
SQL Server transport requires SELECT permission when sending to remote queues

### ISSUE THREAD

**Issue by mauroservienti** [Particular/NServiceBus.SqlServer#1196]
Forum thread:
- https://discuss.particular.net/t/nservicebus-8-needs-select-permission-to-send-messages/3496

### Symptoms

Endpoints sending messages to remote queues fail if the endpoint does not have `SELECT` permissions on the remote queue table. `INSERT` permissions are not enough. The endpoint send/public operations fail with an exception similar to:

```
System.Exception: Failed to send message to [TargetEndpointDB].[TargetEndpointSchema].[TargetEndpointQueue]
 ---> System.Data.SqlClient.SqlException (0x80131904): The SELECT permission was denied on the object 'TargetEndpointQueue', database 'TargetEndpointDB', schema 'TargetEndpointSchema'.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.CompleteAsyncExecuteReader()
   at System.Data.SqlClient.SqlCommand.InternalEndExecuteReader(IAsyncResult asyncResult, String endMethod)
   at System.Data.SqlClient.SqlCommand.EndExecuteReaderInternal(IAsyncResult asyncResult)
   at System.Data.SqlClient.SqlCommand.EndExecuteReader(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location ---
   at NServiceBus.Transport.SqlServer.TableBasedQueue.GetSendCommandText(SqlConnection connection, SqlTransaction transaction, CancellationToken cancellationToken) in /_/src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs:line 151
   at NServiceBus.Transport.SqlServer.TableBasedQueue.SendRawMessage(MessageRow message, SqlConnection connection, SqlTransaction transaction, CancellationToken cancellationToken) in /_/src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs:line 104
ClientConnectionId:713f1bf3-e5ad-43e3-ad85-4200f878f5dc
Error Number:229,State:5,Class:14
   --- End of inner exception stack trace ---
   at NServiceBus.Transport.SqlServer.TableBasedQueue.ThrowFailedToSendException(Exception ex) in /_/src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs:line 180
   at NServiceBus.Transport.SqlServer.TableBasedQueue.SendRawMessage(MessageRow message, SqlConnection connection, SqlTransaction transaction, CancellationToken cancellationToken) in /_/src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs:line 126
   at NServiceBus.Transport.SqlServer.MessageDispatcher.Dispatch(IEnumerable`1 operations, SqlConnection connection, SqlTransaction transaction, CancellationToken cancellationToken) in /_/src/NServiceBus.Transport.SqlServer/Sending/MessageDispatcher.cs:line 147
   at NServiceBus.Transport.SqlServer.MessageDispatcher.DispatchIsolated(SortingResult sortedOperations, TransportTransaction transportTransaction, CancellationToken cancellationToken) in /_/src/NServiceBus.Transport.SqlServer/Sending/MessageDispatcher.cs:line 82
   at NServiceBus.Transport.SqlServer.MessageDispatcher.Dispatch(TransportOperations operations, TransportTransaction transportTransaction, CancellationToken cancellationToken) in /_/src/NServiceBus.Transport.SqlServer/Sending/MessageDispatcher.cs:line 35
   at NServiceBus.TransportReceiveToPhysicalMessageConnector.Invoke(ITransportReceiveContext context, Func`2 next) in /_/src/NServiceBus.Core/Pipeline/Incoming/TransportReceiveToPhysicalMessageConnector.cs:line 63
   at NServiceBus.RetryAcknowledgementBehavior.Invoke(ITransportReceiveContext context, Func`2 next) in /_/src/NServiceBus.Core/ServicePlatform/Retries/RetryAcknowledgementBehavior.cs:line 25
   at NServiceBus.MainPipelineExecutor.Invoke(MessageContext messageContext, CancellationToken cancellationToken) in /_/src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs:line 45
   at NServiceBus.MainPipelineExecutor.Invoke(MessageContext messageContext, CancellationToken cancellationToken) in /_/src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs:line 64
   at NServiceBus.Transport.SqlServer.ProcessStrategy.TryHandleMessage(Message message, TransportTransaction transportTransaction, ContextBag context, CancellationToken cancellationToken) in /_/src/NServiceBus.Transport.SqlServer/Receiving/ProcessStrategy.cs:line 49
   at NServiceBus.Transport.SqlServer.ProcessWithNativeTransaction.TryProcess(Message message, TransportTransaction transportTransaction, ContextBag context, CancellationToken cancellationToken) in /_/src/NServiceBus.Transport.SqlServer/Receiving/ProcessWithNativeTransaction.cs:line 112
Exception details:
	Message ID: 3cd75a97-8441-4e9b-8e90-afce00f12919
	Transport message ID: 50213f42-1e8c-4665-ad8e-a0e9667104a3
	Pipeline canceled: False
```

### Who's affected

Users running endpoints with the `7.x` version of the SQL Server transport without `SELECT` permissions on remote queue tables. 

### Root cause

When sending messages for the first time to a remote queue, the transport performs a query to check the existence of `Reoverable` column:

https://github.com/Particular/NServiceBus.SqlServer/blob/f7c487d60d92a352f3178ecaecfd891c4a0ebd27/src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs#L148-L161

This is done with a `SELECT` statement that requires the permission to be granted.

### Patched version

- 8.1.5
- 7.0.7

<details>
<summary>Original description</summary>

#### Description

As reported in the [public discussion forum](https://discuss.particular.net/t/nservicebus-8-needs-select-permission-to-send-messages/3496) it seems that version 7.x of the transport requires SELECT permission. it's needed at startup to check the presence of the `Recovarable` column:

https://github.com/Particular/NServiceBus.SqlServer/blob/f7c487d60d92a352f3178ecaecfd891c4a0ebd27/src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs#L148-L161

#### Expected behavior

No need for the SELECT permission to send messages.

#### Possible solutions

As mentioned in https://discuss.particular.net/t/nservicebus-8-needs-select-permission-to-send-messages/3496/2 use `SELECT * FROM SYS.COLUMNS` that doesn't require the SELECT permission

</details>

**Comment by MarcWils** [comment]
Any update on this issue? We're about to update a lot of endpoints to version 7 of the transport. We have thight security measures in place. It would be nice if we could keep away from granting SELECT permissions for sending messages.

**Comment by SzymonPobiega** [comment]
I marked it as a candidate for the next release but we can't at this point say when that release it going to happen.

**Comment by blieven1** [comment]
any update on this? this effectivly blocks us to upgrade because we don't want to roll out extra select permissions on everything

**Comment by selism** [comment]
We are running into this issue as well.  We are starting to upgrade our v7 endpoints to v8 and this is limiting our progress.  We have strict security on our central NSB database.  We allow products to send messages to other product's schemas, but we do not allow reading, as a lot of our products work with highly sensitive (wages, sickness, etc...) data.

**Comment by WilliamBZA** [comment]
Fixed by https://github.com/Particular/NServiceBus.SqlServer/pull/1442

Will be released in v8.1.5

**Comment by selism** [comment]
I hope it will also get backported to V7 of NServiceBus.Transport.SqlServer.
Most of our endpoints are .NET Framework applications, so we are stuck on NSB v8 and cannot upgrade to V8 of this package.

**Comment by tmasternak** [comment]
The fixes are out (versions `8.1.5` and `7.0.7` of the package)
